File: backend/src/game/geometry.rs
use super::math::{cross, normalize};
use super::types::Point;
use std::f64::consts::PI;

pub fn tangent_basis(normal: Point) -> (Point, Point) {
    let up = if normal.y.abs() < 0.9 {
        Point {
            x: 0.0,
            y: 1.0,
            z: 0.0,
        }
    } else {
        Point {
            x: 1.0,
            y: 0.0,
            z: 0.0,
        }
    };
    let tangent = normalize(cross(up, normal));
    let bitangent = normalize(cross(normal, tangent));
    (tangent, bitangent)
}

pub fn sample_outline_radius(outline: &[f64], theta: f64) -> f64 {
    if outline.is_empty() {
        return 0.0;
    }
    let total = PI * 2.0;
    let normalized = (theta / total).clamp(0.0, 1.0);
    let idx = normalized * outline.len() as f64;
    let i0 = idx.floor() as usize % outline.len();
    let i1 = (i0 + 1) % outline.len();
    let t = idx - idx.floor();
    outline[i0] * (1.0 - t) + outline[i1] * t
}

File: frontend/src/game/prediction/parity/geometry.ts
import type { Point } from '@game/types'
import { cross, normalize } from '@game/math'

const TAU = Math.PI * 2

export function tangentBasis(normal: Point): { tangent: Point; bitangent: Point } {
  const up = Math.abs(normal.y) < 0.9 ? { x: 0, y: 1, z: 0 } : { x: 1, y: 0, z: 0 }
  const tangent = normalize(cross(up, normal))
  const bitangent = normalize(cross(normal, tangent))
  return { tangent, bitangent }
}

export function sampleOutlineRadius(outline: number[], theta: number): number {
  if (outline.length === 0) return 0
  const normalized = Math.max(0, Math.min(1, theta / TAU))
  const idx = normalized * outline.length
  const floorIdx = Math.floor(idx)
  const i0 = floorIdx % outline.length
  const i1 = (i0 + 1) % outline.length
  const t = idx - floorIdx
  const a = outline[i0] ?? 0
  const b = outline[i1] ?? a
  return a * (1 - t) + b * t
}
